### PR TITLE
fixing out of bound access for cross_entropy

### LINKED ISSUE #1744: `_nll_loss_helper` is trying to index the output of `LogSoftmax` with an `ignore_index` value which is out of bounds!
## 🐛 Bug

_Note: I discovered this while trying to enable cross-entropy-loss through nvFuser.  The repro is purely through Thunder's torch executor._

In the `_nll_loss_helper` that decomposes `cross_entropy` loss, prior to index the output of `LogSoftmax` with the labels that are modified with an `ignore_index` of `-100`.  This index needs to be masked out because it is not in range of the vocabulary and results in an error as follows:

```bash
/opt/pytorch/pytorch/aten/src/ATen/native/cuda/ScatterGatherKernel.cu:144: operator(): block: [1,0,0], thread: [127,0,0] Assertion `idx_dim >= 0 && idx_dim < index_size && "index out of bounds"` failed.
```

The issue is at the following lines:
https://github.com/Lightning-AI/lightning-thunder/blob/40f79727c419d3acf83cd75b51fa867810e23a46/thunder/torch/__init__.py#L5117-L5119

I think the `where` line needs to produce the target indices supplied to `take_along_dim`.

### To Reproduce

```bash
git clone https://github.com/kevinstephano/thunder_model_blocks.git
cd thunder_model_blocks
python hf_qwen2/qwen2_loss.py --execs Thunder-Torch
```

cc @tfogal

### PR DESCRIPTION
**Pull request by jjsjann123**
Fixes #1744 

Just a temporary solution

### CODE REVIEW

File: thunder/torch/__init__.py [hunk 1]
```diff
@@ -5111,12 +5111,18 @@ def _nll_loss_helper(
         bcast_weight = reshape(weight, [num_class] + [1 for _ in range(2, a.ndim)])
         out = out * bcast_weight
 
+    assert isinstance(ignore_index, Number)
```
[the remark this answers is not in the document]
Comment (jjsjann123) [reply thunder/torch/__init__.py:5114]: For today with constant constraints, I think that's right. But then we could have the same out of bound access issue with NumberProxy for ignore_index.
If we do need to support that, I think for now we need to use `< 0 or >= num_class` to handle it.

[the remark this answers is not in the document]
Comment (jjsjann123) [reply thunder/torch/__init__.py]: BTW, added the support.

For some reason thunderfx does indeed hit give `ignore_index` as NumberProxy instead, you probably know this better than I do.

**Comment by jjsjann123** [comment]
This PR is good for review now. @IvanYashchuk @mruberry

File: thunder/torch/__init__.py [hunk 2]
```diff
@@ -5092,31 +5092,50 @@ def _nll_loss_helper(
...
+    else:
+        # scenario 2 in handling `ignore_index` parameter
+        # i. cap the ignore_index to be num_class
+        selected_target_mask = bcast_target != ignore_index
+        bcast_target = where(selected_target_mask, bcast_target, num_class)
+        # ii. pad the logits to have an all-zero entry at index num_class
+        padding = [(0, 0, 0)] * out.ndim
+        padding[class_dim] = (0, 1, 0)
+        padded_out = clang.pad(out, utils.const_as(0, out.dtype), padding)
+        # with the cap and pad, all ignore_index entries would be all zero, mimicing the skip behabior
```
Comment (IvanYashchuk) [inline thunder/torch/__init__.py:5137]:
```suggestion
        # with the cap and pad, all ignore_index entries would be all zero, mimicking the skip behavior
```

**Review by IvanYashchuk — APPROVED** [review]

**Review by t-vi — APPROVED** [review]
Thank you @jjsjann123 @IvanYashchuk

File: thunder/torch/__init__.py (hunk 2 above)
Comment (jjsjann123) [reply thunder/torch/__init__.py]: :saluting_face:
